content/ui-patterns/saving.mdx

```diff
@@ -0,0 +1,63 @@
+
+### Optimistic update
+For interactions where the entity already exists and we merely need to update a field on that object, we should initiate an async server request using `remoteForm`, immediately display the desired changes as if they've already happened, and upon response from the server, either cache the result of a successful change to save loading time later or present a clear, concise error message if an error has occurred. If the change may not be obviously visible in the UI the user is currently viewing, we should trigger a [Toast](https://primer.style/css/components/toasts) to let the user know that the change occurred successfully. 
+
+```javascript
+
+<!-- feature-preview-trigger.ts -->
+
+import {fire, on} from 'delegated-events'
+import {getAttribute, query, querySelectorAll} from '@github/query-selector'
+import {dialog} from '../details-dialog'
+import {fetchSafeDocumentFragment} from '../fetch'
+import {remoteForm} from '@github/remote-form'
+
+...
+
+on('click', '[data-feature-preview-trigger-url]', async event => {
+  const trigger = event.currentTarget
+  const url = getAttribute(trigger, 'data-feature-preview-trigger-url')
+
+  const newDialog = await dialog({
+    content: fetchSafeDocumentFragment(document, url),
+    dialogClass: 'feature-preview-dialog'
+  })
+
+  const indicators = querySelectorAll(document, '.js-feature-preview-indicator')
+  for (const indicator of indicators) {
+    indicator.hidden = true
+  }
+})
+
+remoteForm('.js-feature-preview-unenroll', async (form, wants) => {
+  await wants.text()
+  const feature = query(form, '.js-feature-preview-slug', HTMLInputElement).value
+  fire(form, `feature-preview-unenroll:${feature}`)
+})
+
+remoteForm('.js-feature-preview-enroll', async (form, wants) => {
+  await wants.text()
+  const feature = query(form, '.js-feature-preview-slug', HTMLInputElement).value
+  fire(form, `feature-preview-enroll:${feature}`)
+})
+
+```
+  
+In a few specific cases, such as updating notification preferences, we should use the `data-disable-with` property to indicate that the response is still being sent to signal to the user that they shouldn't navigate away from the current page which may cause the request to fail silently in the event of an error.
+
+```html
+  <input type="submit" data-disable-with="Submitting...">Submit</input>
```

Comment:
When is it ok to use a different string in `data-disable-with`? If the request is fast, the effect is a bit harsh.

![CleanShot 2020-07-16 at 17 32 24](https://user-images.githubusercontent.com/3992/87735573-b982e900-c78a-11ea-9ebf-3831f574120b.gif)

Reply:
That's not something we can control though, so we should support the worst-case scenario, right?

Reply:
I believe `data-disable-with` always requires a string, but it probably makes sense to always use the present-tense form of the initial verb?

Reply:
We definitely can't control the speed of the request, but we can control the copy.

It seems that by recommending a copy change on `data-disable-with` we are opening the door to a lot of fast UI copy jumps.

Reply:
But then, if it's a long pause and there's no change, that seems worse, right? We could stick to disabling the button, but even that doesn't necessarily declare that something is happening...

Reply:
Could we swap in a loading indicator instead of text perhaps?